[documentation] template for new component issue

codeboten opened the pull request:
Adding a template to use as an experiment in this repository. The workflow is expecting users to open a new component issue when proposing an issue, to reduce unnecessary work both for contributors and maintainers.

Once the template is approved, it can be copied over to the core repository and linked from the docs as well.

Commit:
[documentation] template for new component issue

Adding a template to use as an experiment in this repository. The workflow is expecting users to open a new component issue when proposing an issue, to reduce unnecessary work both for contributors and maintainers.

bogdandrutu commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] Add component to `versions.yaml`
+- [ ] `CODEOWNERS` with at least the sponsor and the component's author
+- [ ] internal/components tests
+- [ ] Add component to `go.mod` at the root of the repository and at `cmd/configschema`
```
This is done automatically (when you link to `internal/components`), they need only the replace statement correct?

bogdandrutu commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] internal/components tests
+- [ ] Add component to `go.mod` at the root of the repository and at `cmd/configschema`
+- [ ] Enable Dependabot for your component by running `make gendependabot`.
+- [ ] The skeleton of the component, potentially using the helpers
```
Is this something decided in the SIG meeting to not be a step, since we accept one large PR? I am trying to say that this is normal, not sure we need to link it here as a step.

bogdandrutu commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+
+**Checklist**
+
+The following is a checklist for tracking the work required when adding a new component:
```
Should we sort it somehow?

bogdandrutu commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+The following is a checklist for tracking the work required when adding a new component:
+
+- [ ] Identify a sponsor
+- [ ] Document README.md with some basic information about your component. It doesn't have to be complete, but should contain at least:
+  - [ ] The purpose of the component
+  - [ ] Use-cases
```
Is this not what we should ask in the issue template, since the first PR from the current process is no longer required?

jpkrohling replied:
+1, the original "checklist" mentioned a readme file with these two items, but they should now be included in the issue, as there's no initial PR.

bogdandrutu commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] Document README.md with some basic information about your component. It doesn't have to be complete, but should contain at least:
+  - [ ] The purpose of the component
+  - [ ] Use-cases
+- [ ] A complete config.go, which will show how your component will look like to users
```
Same, this is the example config from your template? when they write the code they definitely need a config. Maybe a separate section we can explain how to structure a component: config (embed the settings, bla bla); factory (use the helpers, bla bla), etc.

bogdandrutu commented:
@codeboten see also https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#adding-new-components

tigrannajaryan commented:
Do we need this? Is there any evidence that users often create issues of this kind? Check existing issues perhaps to show that this is a top request type?

codeboten commented:
> Do we need this? Is there any evidence that users often create issues of this kind? Check existing issues perhaps to show that this is a top request type?

This was proposed in the SIG meeting on Jan 5 to reduce friction for reviewers and contributors. The thinking here is that this would replace the need to have a first PR when adding a component. Instead the issue containing the same level of information as that initial PR would suffice.

It's related to https://github.com/open-telemetry/opentelemetry-collector/issues/4630

jpkrohling commented:
I just updated #6926 with a comment based on my recollection of what we agreed during the latest SIG call: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/6926#issuecomment-1009153700

The major difference is that it just mentioned the issue that has to be created, what the initial PR should contain, and the step about the -releases repo. I think we might then need two templates, one for the initial issue and one for the PR.

mx-psi commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,41 @@
+- [ ] Add component to `go.mod` at the root of the repository and at `cmd/configschema`
+- [ ] Enable Dependabot for your component by running `make gendependabot`.
+- [ ] The skeleton of the component, potentially using the helpers
+- [ ] Add Makefile
```
This may be confusing to people ("what should the Makefile contents be?") so maybe we can clarify this (e.g. "copy from template here")

Commit:
remove checklist, update content with links to examples

codeboten commented:
Removed the checklist, which can be added as a PR checklist. Updated the template with some links to example content and updated the contributing document to point to creating a new issue. @bogdandrutu @jpkrohling @mx-psi @tigrannajaryan PTAL

tigrannajaryan commented on CONTRIBUTING.md:
```diff
@@ -6,6 +6,13 @@ before you begin your work.
 
+Before any code is written, open an issue providing the following information:
+
+    * Who's the sponsor for your component. A sponsor is an approver who will be in charge of being the official reviewer of the code. For vendor-specific components, it's good to have a volunteer sponsor. If you can't find one, we'll assign one in a round-robin fashion. For non-vendor specific components, having a sponsor means that your use case has been validated.
```
Is the sponsor only responsible for the review or also for subsequent maintenance?

codeboten replied:
Thanks for the callout, i've added a note about the sponsor becoming a code owner as well.

Commit:
update based on feedback

codeboten commented on CONTRIBUTING.md:
```diff
@@ -6,6 +6,13 @@ before you begin your work.
+
+    * Who's the sponsor for your component. A sponsor is an approver who will be in charge of being the official reviewer of the code. For vendor-specific components, it's good to have a volunteer sponsor. If you can't find one, we'll assign one in a round-robin fashion. For non-vendor specific components, having a sponsor means that your use case has been validated.
+    * Some information about your component, such as the reasoning behind it, use-cases, telemetry data types supported, and anything else you think it's relevant for us to make a decision about accepting the component
+    * The configuration options your component will accept. This will help us understand what it does and have an idea of how the implementation might look like
```
Fixed, i fixed the formatting earlier but hadn't saved it before submitting the change.

JamieDanielson commented on .github/ISSUE_TEMPLATE/new_component.md:
```diff
@@ -0,0 +1,24 @@
+
+**The purpose and use-cases of the new component**
+
+*This information can be used later on to populate the README for the component. See an example overview [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/awsecscontainermetricsreceiver#overview).*
```
```suggestion
<!--
*This information can be used later on to populate the README for the component. See an example overview [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/awsecscontainermetricsreceiver#overview).*
-->
```

Any reason to not make these commented blocks in the issue template? Might help ensure submitters keep the header (consider making a `##`) and remove this block to reduce noice in submission

codeboten replied:
Good call, done!

jpkrohling commented on CONTRIBUTING.md:
```diff
@@ -6,6 +6,13 @@ before you begin your work.
 
+**Before** any code is written, [open an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=new+component&template=new_component.md&title=New%20component) providing the following information:
+
+* Who's the sponsor for your component. A sponsor is an approver who will be in charge of being the official reviewer of the code and become a code owner for the component. For vendor-specific components, it's good to have a volunteer sponsor. If you can't find one, we'll assign one in a round-robin fashion. For non-vendor specific components, having a sponsor means that your use case has been va ...
```
Sometimes, the idea for a component circulates via Slack first. But I agree that the typical case would probably be that sponsors have to be found after the idea is exposed via the initial issue.

So, add a new statement to this phrase, like:

> If there are no sponsors yet for the component, it's fine: use the issue as a means to try to find a sponsor for your component.

codeboten replied:
Added.

Commit:
update based on review feedback

djaglowski commented on CONTRIBUTING.md:
```diff
@@ -6,6 +6,13 @@ before you begin your work.
+**Before** any code is written, [open an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=new+component&template=new_component.md&title=New%20component) providing the following information:
+
+* Who's the sponsor for your component. A sponsor is an approver who will be in charge of being the official reviewer of the code and become a code owner for the component. For vendor-specific components, it's good to have a volunteer sponsor. If you can't find one, we'll assign one in a round-robin fashion. For non-vendor specific components, having a sponsor means that your use case has been va ...
+* Some information about your component, such as the reasoning behind it, use-cases, telemetry data types supported, and anything else you think it's relevant for us to make a decision about accepting the component
+* The configuration options your component will accept. This will help us understand what it does and have an idea of how the implementation might look like
```
```suggestion
* Some information about your component, such as the reasoning behind it, use-cases, telemetry data types supported, and anything else you think is relevant for us to make a decision about accepting the component.
* The configuration options your component will accept. This will help us understand what it does and have an idea of how the implementation might look like.
```

djaglowski commented on CONTRIBUTING.md:
```diff
@@ -6,6 +6,13 @@ before you begin your work.
+**Before** any code is written, [open an issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=new+component&template=new_component.md&title=New%20component) providing the following information:
+
+* Who's the sponsor for your component. A sponsor is an approver who will be in charge of being the official reviewer of the code and become a code owner for the component. For vendor-specific components, it's good to have a volunteer sponsor. If you can't find one, we'll assign one in a round-robin fashion. For non-vendor specific components, having a sponsor means that your use case has been va ...
+* Some information about your component, such as the reasoning behind it, use-cases, telemetry data types supported, and anything else you think it's relevant for us to make a decision about accepting the component
```
> for us to make a decision about accepting the component

Once the issue is opened, what is the process for making a decision? How is the decision indicated?

codeboten replied:
What are your thoughts? The best I can think of is approvers/maintainers commenting on the issue that a decision has been made to approve/reject a proposal.

bogdandrutu replied:
Labels? "Proposed Component", "Approved Component", and after X weeks close the issue as no decision made => no interest in supporting this?

djaglowski replied:
> Labels? "Proposed Component", "Approved Component", and after X weeks close the issue as no decision made => no interest in supporting this?

I like this. The "Sponsor Needed" label can be applied until a decision is made.

Commit:
Update CONTRIBUTING.md

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>
